rfd/00TBD-webui-paginate-filter.md

```diff
@@ -0,0 +1,335 @@
+
+We can branch off into two functions with current `ListResources` based on if sorting was requested. Sorting will be disabled for `tsh`, so that `tsh` performance will not be affected:
+
+  - `listResources` (keeps current behavior)
+  - `listResourcesWithSort`
+
+`listResourcesWithSort` will:
+
+1. retrieve the entire list of resources ie: `GetNodes` (operating on cache)
+1. apply filters (rbac > label + search)
+1. sort
+1. paginate
+
+Similarly to how we did on the web UI, we have to decide which fields of a resource will be sortable. There will be a sort data type that stores the sort column (field const of a resource) and the sort direction (asc or desc). Custom sort function will be used for each resource type. The column name will be defined as const strings that will be used by both client and server side:
+
+**Example:**
+
+| Server Columns       | Const String  |
+|----------------------|---------------|
+| Server.Spec.Hostname | "colHostname" |
+| Server.Spec.Addr     | "colAddr"     |
+
+<br />
+
+| App Server Columns       | Const String     |
+|--------------------------|------------------|
+| App.Metadata.Name        | "colName"        |
+| App.Metadata.Description | "colDescription" |
+| App.Spec.PublicAddr      | "colPublicAddr"  |
+
+<br />
+
+| Db Server Columns         | Const String     |
+|---------------------------|------------------|
+| DB.Metadata.Name          | "colName"        |
+| DB.Metadata.Description   | "colDescription" |
+| DB.Spec.(AWS,Azure...)... | "colType"        |
+
+<br />
+
+| Kube Cluster Columns | Const String |
+|----------------------|--------------|
+| KubeCluster.Name     | "colName"    |
+
+<br/>
+
+### Filter: Label + Search
+
+Currently, only node resource has the capability to be matched against labels (other resources are [pending](https://github.com/gravitational/teleport/pull/9096#issuecomment-989291135)). The matching needs to be extended to also support filtering by search values and support for `NOT` and `OR` label matching.
+
+In order to expand filtering to support NOT and OR, we can use the `vulcand/predicate` library.
+
+Assuming the query string is in the form the `vulcand/predicate` library expects:
+1. Create a custom predicate parser that on parse `query string` ouputs an expression tree (similar to [newParserForIdentifierSubcondition](https://github.com/gravitational/teleport/blob/master/lib/services/parser.go#L416))
+1. Feed this expression tree to [ToFieldsCondition](https://github.com/gravitational/teleport/blob/master/lib/utils/fields.go#L100) that returns boolean function to apply on fields.
+1. For each row, create a `Fields` map that contains all the field values we want to test against the conditions created from previous step. [(example of a similar usage done in teleport)](https://github.com/gravitational/teleport/blob/master/lib/utils/fields_test.go#L52)
```

Comment:
The idea of "transpiling" the query to another query language feels a bit brittle/hacky to me.  If this is really the direction we need to go, so be it.  Check out the alternative mentioned in my top-level comment though.  I think we could save a lot of potential headaches by simplifying our model just a bit.